src/ray/common/asio/asio_chaos.cc

```diff
@@ -0,0 +1,136 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/common/asio/asio_chaos.h"
+
+#include <cstdlib>
+#include <string_view>
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/strings/numbers.h"
+#include "absl/strings/str_split.h"
+#include "ray/common/ray_config.h"
+#include "ray/util/logging.h"
+
+namespace ray {
+namespace asio {
+namespace testing {
+
+namespace {
+/*
+  DelayManager is a simple chaos testing framework. Before starting ray, users
+  should set up os environment to use this feature for testing purposes.
+
+  To use this, simply do
+      export RAY_testing_asio_delay_us="method1=10:10,method2=20:100"
+
+   The delay is a random number between the left and right of colon. If method equals '*',
+   it will apply to all methods.
+
+   Please check warnings to make sure delay is on.
+*/
+class DelayManager {
+ public:
+  DelayManager() { Init(); }
+
+  int64_t GetMethodDelay(const std::string &name) const {
+    auto it = delay_.find(name);
+    if (it == delay_.end()) {
+      return GenRandomDelay(random_delay_us_.first, random_delay_us_.second);
+    }
+    int64_t actual_delay = GenRandomDelay(it->second.first, it->second.second);
+    if (actual_delay != 0) {
+      RAY_LOG_EVERY_N(ERROR, 1000)
+          << "Delaying method " << name << " for " << actual_delay << "us";
+    }
+    return actual_delay;
+  }
+
+  void Init() {
+    delay_.clear();
+    random_delay_us_ = {0, 0};
+    auto delay_env = RayConfig::instance().testing_asio_delay_us();
+    if (delay_env.empty()) {
+      return;
+    }
+    RAY_LOG(ERROR) << "RAY_testing_asio_delay_us is set to " << delay_env;
+    std::vector<std::string_view> items = absl::StrSplit(delay_env, ",");
+    for (const auto &item : items) {
+      ParseItem(item);
+    }
+  }
+
+ private:
+  // method1=min:max
+  void ParseItem(std::string_view val) {
+    std::vector<std::string_view> item_val = absl::StrSplit(val, "=");
+    if (item_val.size() != 2) {
+      RAY_LOG(ERROR) << "Error in syntax: " << val
```

Comment:
```suggestion
      RAY_LOG(FATAL) << "Error in syntax: " << val
```

Let's just check fail here, to avoid tests accidentally passing with bad syntax.